src/Build/Evaluation/Expander.cs

```diff
@@ -2747,6 +2832,111 @@ internal static IEnumerable<KeyValuePair<string, S>> ExpandQuotedExpressionFunct
+                /// </summary>
+                private static OneOrMultipleMetadataMatches GetQuotedExpressionMatches(string quotedExpressionFunction, IElementLocation elementLocation)
+                {
+                    // Start with fast paths to avoid any allocations.
+                    if (TryGetCachedMetadataMatch(quotedExpressionFunction, out string cachedName)
+                        || s_itemSpecModifiers.TryGetValue(quotedExpressionFunction, out cachedName))
+                    {
+                        return new OneOrMultipleMetadataMatches(cachedName);
+                    }
+
+                    Match match = RegularExpressions.ItemMetadataRegex.Match(quotedExpressionFunction);
+
+                    // GroupCollection + Groups are the most expensive source of allocations here, so we want to return
+                    // before ever accessing the property. Simply accessing it will trigger the full collection
+                    // allocation, so we avoid it unless absolutely necessary.
+                    // Unfortunately even .NET Core does not have a struct-based Group enumerator at this point.
+                    if (!match.Success)
+                    {
+                        return new OneOrMultipleMetadataMatches(string.Empty);
+                    }
+                    else if (s_itemSpecModifiers.TryGetValue(match.Value, out cachedName))
+                    {
+                        return new OneOrMultipleMetadataMatches(quotedExpressionFunction, match, cachedName);
+                    }
+
+                    // From here will either return:
+                    // 1. A single match.
+                    // 2. A list of multiple matches.
+                    GroupCollection groupCollection = match.Groups;
+                    string name = groupCollection[RegularExpressions.NameGroup].Value;
+
+                    ProjectErrorUtilities.VerifyThrowInvalidProject(groupCollection[RegularExpressions.ItemSpecificationGroup].Length == 0, elementLocation, "QualifiedMetadataInTransformNotAllowed", match.Value, name);
+
+                    Match firstMatch = match;
+                    match = firstMatch.NextMatch();
+
+                    if (!match.Success)
+                    {
+                        OneOrMultipleMetadataMatches singleMatch = new(quotedExpressionFunction, firstMatch, name);
+
+                        if (singleMatch.Type == MetadataMatchType.ExactString)
+                        {
+                            // Only cache full string matches.
+                            s_lastParsedQuotedExpression = name;
+                        }
+
+                        return singleMatch;
+                    }
+
+                    // Now we run the full loop.
+                    // This is a very hot path, so we avoid allocating this until after we know there are multiple matches.
+                    List<MetadataMatch> multipleMatches = [new MetadataMatch(firstMatch, name)];
```

Comment:
Can you comment an example expression that falls to this case? It makes sense from the code but I'm having a hard time backing into what the XML looked like to get here.

Reply:
Mostly stuff like `%(Filename)%(Extension)`, but you get ones like:

`Microsoft.NET.GenerateAssemblyInfo.targets` (mix of known modifiers + other)
```xml
<Hash ItemsToHash="@(AssemblyAttribute->'%(Identity)%(_Parameter1)%(_Parameter2)%(_Parameter3)%(_Parameter4)%(_Parameter5)%(_Parameter6)%(_Parameter7)%(_Parameter8)')">
```

`Microsoft.NET.Sdk.BeforeCommon.targets` (string interpolation)
```xml
<Target Name="GenerateNETCompatibleDefineConstants">
...
    <_ImplicitDefineConstant Include="@(_FormattedCompatibleFrameworkVersions->'NETCOREAPP%(Identity)_OR_GREATER'->Replace('.', '_'))" />
```